### PR TITLE
Change the builtin markdown filter to use the markdown renderer from the renderer crate (fix #943)

### LINKED ISSUE #943: How to get syntax highlighting in shortcodes?
# Bug Report
Syntax highlighting is not applied within a shortcode block properly.

Suppose we want to create a shortcut to mimic the details/summary collapsible section in markdown that GitHub and others use.

```html
<details>
    <summary>{{ summary }}</summary>

{{ body | markdown(inline=true) | safe }}

</details>
```

Then, if we try and apply the shortcode as follows, with syntax highlighting enabled.

````
{% details(summary="Click Here") %}
```xml
<item>
  <title>Some title</title>
  <link>www.example.com</link>
</item>
```
{% end %}

```xml
<item>
  <title>Some title</title>
  <link>www.example.com</link>
</item>
```
````

## Environment
Ubuntu 18.04

Zola version:
0.9.0

## Expected Behavior
Tell us what should have happened.

We would expect that syntax highlighting is applied for both blocks of code.

## Current Behavior
Tell us what happens instead of the expected behavior.

Instead, we see that the they are not rendered the same:
screenshot

With the corresponding source:

```html
<details>
    <summary>Click Here</summary>
<pre><code class="language-xml">&lt;item&gt;
  &lt;title&gt;Some title&lt;/title&gt;
  &lt;link&gt;www.example.com&lt;/link&gt;
&lt;/item&gt;
</code></pre>
</details>
<pre style="background-color:#fdf6e3;">
<span style="color:#93a1a1;">&lt;</span><span style="color:#268bd2;">item</span><span style="color:#93a1a1;">&gt;
  &lt;</span><span style="color:#268bd2;">title</span><span style="color:#93a1a1;">&gt;</span><span style="color:#657b83;">Some title</span><span style="color:#93a1a1;">&lt;/</span><span style="color:#268bd2;">title</span><span style="color:#93a1a1;">&gt;
  &lt;</span><span style="color:#268bd2;">link</span><span style="color:#93a1a1;">&gt;</span><span style="color:#657b83;">www.example.com</span><span style="color:#93a1a1;">&lt;/</span><span style="color:#268bd2;">link</span><span style="color:#93a1a1;">&gt;
&lt;/</span><span style="color:#268bd2;">item</span><span style="color:#93a1a1;">&gt;
</span></pre>
```

## Step to reproduce
Please provide the steps to reproduce the issue.

Use the shortcode and markdown file provided in the description.

### PR DESCRIPTION
**Pull request by uggla**
Sanity check:

* [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/getzola/zola/pulls) for the same update/change?

## Code changes
(Delete or ignore this section for documentation changes)

* [X] Are you doing the PR on the `next` branch?

If the change is a new feature or adding to/changing an existing one:

* [X] Have you created/updated the relevant documentation page(s)?

Explanations:
I have changed the builtin markdown filter code to use the markdown renderer from the renderer crate
I have extended the filter arguments to pass highlight and highlight_theme values to the filter. I did that because getting the values from the configuration within the filter looks far more complex.
The is a slight difference in the rendered html, that I will point out in a PR comment.

This PR only addresses the major problem and not the short code nesting issue reported by @mattly.

### CODE REVIEW

File: components/templates/src/filters.rs [hunk 1]
```diff
@@ -64,7 +73,17 @@ mod tests {
     fn markdown_filter() {
         let result = markdown(&to_value(&"# Hey").unwrap(), &HashMap::new());
         assert!(result.is_ok());
-        assert_eq!(result.unwrap(), to_value(&"<h1>Hey</h1>\n").unwrap());
+        // zola render adds an id to the h1 tag
+        assert_eq!(result.unwrap(), to_value(&"<h1 id=\"hey\">Hey</h1>\n").unwrap());
```
Comment (uggla) [inline components/templates/src/filters.rs:77]: Slight rendering change, the renderer crate inserts an id to the h1 tag. I considered this ok, and changed the test.

**Comment by Keats** [comment]
It's actually pretty easy to pass the config to a filter!
Have a look at `GetUrl` in `components/templates/src/global_fns/mod.rs`, it works the same way with filters

**Comment by uggla** [comment]
> It's actually pretty easy to pass the config to a filter!
> Have a look at `GetUrl` in `components/templates/src/global_fns/mod.rs`, it works the same way with filters

@Keats, I may miss something here. Although I spent some time looking at the code.
I have looked at the `GetUrl` method, if I understand well the site.config is passed when the function is registered this looks to be in `tpls.rs` file in the `register_early_global_fns`. So we can use the live configuration into `GetUrl` and grab the data we need.

However it looks a bit different regarding filters. The `markdown` filter is registered as an extension of tera filters this is triggered by the `tera.extend(&ZOLA_TERA)` function into `tpls.rs`, then it goes into `templates/src/lib.rs` and the function `tera.register_filter("markdown", filters::markdown);`
To my mind, I don't see how the live configuration can be passed and used in the markdown filter, because it seems not passed to the tera module that will register it.
Of course, I may have not understand something and there is a simple trick to pass it differently ? So a bit of help will be appreciated.

**Comment by Keats** [comment]
Have a look at `register_early_global_fns`, it works exactly the same way for filters. I can update that PR later if it's still unclear.

**Comment by uggla** [comment]
> it works exactly the same way for filters.

I have looked at it, but the mechanism for filters looks different to me as explained above.
The Filter trait is defined with only value (name) and args (argument pass to the filter), so I think I can not add one more argument to the markdown filter.

Yes still unclear sorry. :)

**Comment by Keats** [comment]
I've done it in https://github.com/getzola/zola/pull/1189/commits/b5ddae023402e83dc1554c511e86dd1406e741ba if you're curious.

**Comment by uggla** [comment]
Hello @Keats, happy to have news from your side after some time. Hoping you are fine despite the sanitary situation.
Gving a quick glance  to the solution it looks less trivial that I expected. However I will look deeply at it to better understand.
Anyway thanks for the solution and sorry not to be able to help more on the topic.
Feel free to tell me if I can help on something simple.

**Comment by Keats** [comment]
Don't worry about it! Ping me if there's something you don't understand in that commit.

As for issues there is https://github.com/getzola/zola/issues/1249 , https://github.com/getzola/zola/issues/1244 and https://github.com/getzola/zola/issues/1247 are the 3 issues I want to tackle next, feel free to take one or several of them.

**Pull request closed without merging**